packages/sdk-js/src/DidHelpers/checkResult.ts

```diff
@@ -0,0 +1,241 @@
+    case 'Retracted':
+      status = 'unknown'
+      error = new Error(result.status.type)
+      break
+    default:
+      status = 'unknown'
+      error = new Error(`unknown tx status variant ${result.status.type}`)
+  }
+
+  return {
+    status,
+    error,
+    blockHash,
+    blockNumber,
+  }
+}
+
+async function resolveBlockAndEvents(
+  result: { blockHash: HexString; txHash: HexString },
+  api: ApiPromise
+): Promise<{
+  blockNumber: BigInt
+  blockHash: HexString
+  txEvents: EventRecord[]
+}> {
+  const txHashHash = api.createType('Hash', result.blockHash)
+  const {
+    block: { block },
+    events,
+  } = await api.derive.tx.events(txHashHash)
+  const blockNumber = block.header.number.toBigInt()
+  const { blockHash } = result
+  const txIndex = block.extrinsics.findIndex((tx) => tx.hash.eq(result.txHash))
+  const txEvents = events.filter(
+    ({ phase }) => phase.isApplyExtrinsic && phase.asApplyExtrinsic.eqn(txIndex)
+  )
+  return { blockNumber, blockHash, txEvents }
+}
+
+function checkIfEventsMatch(
+  expectedEvents: Array<{ section: string; method: string }>,
+  txEvents: EventRecord[]
+): boolean {
+  return expectedEvents.every(({ section, method }) =>
+    txEvents.some(
+      ({ event }) =>
+        event.section.toLowerCase() === section.toLowerCase() &&
+        event.method.toLowerCase() === method.toLowerCase()
+    )
+  )
+}
+
+export async function checkResultImpl(
+  result: { blockHash: HexString; txHash: HexString } | SubmittableResultValue,
+  api: ApiPromise,
+  expectedEvents: Array<{ section: string; method: string }> = [],
+  did: Did,
+  signersOrKeys: SharedArguments['signers']
+): Promise<TransactionResult> {
+  let txEvents: EventRecord[] = []
+  let status: TransactionResult['status'] | undefined
+  let error: Error | undefined
+  let blockHash: HexString | undefined
+  let blockNumber: BigInt | undefined
+
+  // Case where `SubmittableResultValue` is provided.
+  if ('status' in result) {
+    txEvents = result.events ?? []
+    ;({ status, blockHash, blockNumber, error } = checkStatus(result))
+  }
+  // Case where block hash and tx hash are provided.
+  else if (
+    'blockHash' in result &&
+    'txHash' in result &&
+    typeof result.blockHash === 'string' &&
+    typeof result.txHash === 'string'
+  ) {
+    // Set blockNumber, blockHash, and the transactionEvents.
+    ;({ blockNumber, blockHash, txEvents } = await resolveBlockAndEvents(
+      result,
+      api
+    ))
+  }
+  // Type of the `result` argument is invalid.
+  else {
+    status = 'unknown'
+    error = new Error('missing blockHash and/or txHash')
+  }
+
+  if (typeof status !== 'string') {
+    const eventError = checkEventsForErrors(api, txEvents)
+    if (eventError !== undefined) {
+      error = eventError
+    }
+
+    const eventMatch = checkIfEventsMatch(expectedEvents, txEvents)
+    const isSuccess = !error && eventMatch
+
+    status = isSuccess ? 'confirmed' : 'failed'
+    if (!isSuccess && !error) {
```

Comment:
```suggestion
    if (!isSuccess && !eventMatch) {
```
why not be more explicit?

Reply:
5fa8bed